galaxy/main/migrations/0062_namespace_refactor.py

```diff
@@ -0,0 +1,189 @@
+                ('provider', models.ForeignKey(related_name='provider_namespaces', verbose_name=b'Provider', to='main.Provider', null=True)),
+            ],
+            options={
+                'ordering': ('provider', 'name',),
+                'unique_together': set([('provider', 'name'), ('namespace', 'provider', 'name')]),
+            },
+            bases=(models.Model, galaxy.main.mixins.DirtyMixin),
+        ),
+
+        migrations.RunSQL(
+            sql=COPY_NAMESPACE_DATA,
+            reverse_sql=migrations.RunSQL.noop
+        ),
+
+        migrations.DeleteModel(
+            name='Namespace',
+        ),
+
+        migrations.CreateModel(
+            name='Namespace',
+            fields=[
+                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
+                ('created', models.DateTimeField(auto_now_add=True)),
+                ('modified', models.DateTimeField(auto_now=True)),
+                ('description', galaxy.main.fields.TruncatingCharField(default=b'', max_length=255, blank=True)),
+                ('active', models.BooleanField(default=True, db_index=True)),
+                ('name', models.CharField(unique=True, max_length=512, db_index=True)),
+                ('original_name', models.CharField(max_length=512)),
+                ('avatar_url', models.CharField(max_length=256, null=True, verbose_name=b'Avatar URL', blank=True)),
+                ('location', models.CharField(max_length=256, null=True, verbose_name=b'Location', blank=True)),
+                ('company', models.CharField(max_length=256, null=True, verbose_name=b'Company Name', blank=True)),
+                ('email', models.CharField(max_length=256, null=True, verbose_name=b'Email Address', blank=True)),
+                ('html_url', models.CharField(max_length=256, null=True, verbose_name=b'Web Site URL', blank=True)),
+                ('owners', models.ManyToManyField(related_name='+', to=settings.AUTH_USER_MODEL)),
+            ],
+            options={
+                'ordering': ('name',),
+            },
+            bases=(models.Model, galaxy.main.mixins.DirtyMixin),
+        ),
+
+        migrations.RunSQL(
+            sql=ADD_REPO_GITHUB_USERS,
+            reverse_sql=migrations.RunSQL.noop
+        ),
+
+        migrations.RunSQL(
+            sql=ADD_ROLE_NAMESPACE,
+            reverse_sql=migrations.RunSQL.noop
+        ),
+
+        migrations.RunSQL(
+            sql=NAMESPACE_FROM_PROVIDER_NAMESPACE,
+            reverse_sql=migrations.RunSQL.noop
+        ),
+        migrations.RunSQL(
+            sql=ADD_GITHUB_PROVIDER,
+            reverse_sql=migrations.RunSQL.noop
+        ),
+        migrations.RunSQL(
+            sql=SET_PROVIDER_NAMESPACE_FK,
+            reverse_sql=migrations.RunSQL.noop
+        ),
+        migrations.RunSQL(
+            sql=NAMESPACE_OWNERS,
+            reverse_sql=migrations.RunSQL.noop
```

Comment:
Have you tested migration downgrade? If this migration is irreversible by design, please don't use `migrations.RunSQL.noop`, so it will fail if somebody try to downgrade it.

Reply:
I honestly didn't put any thought into making it reversible. I would say that it's 'irreversible by design'. I didn't know the 'noop' would raise an error. Will take that out.